### ISSUE TITLE
group by showprogress too chatty

### ISSUE THREAD

**Issue by ethanbsmith** [Rdatatable/data.table#6514]
hate to nitpick on such a great package.

i'm seeing things like this: `Processed 5332 groups out of 5332. 100% done. Time elapsed: 3s. ETA: 0s.`

progress on a 3 second op is getting annoying. maybe up the threshold significantly (like at least 30 seconds) or have the threshold as a configurable option?

**Comment by joshhwuu** [comment]
The current threshold was made to be consistent with `fread` and `fwrite` (2 seconds IIRC), but since `group by` is probably more common, an option to configure the threshold may be a good idea (maybe even globally for all `showProgress`?

**Comment by tdhock** [comment]
you can turn it off using the showProgress argument/option, from ?data.table
```
       showProgress = getOption("datatable.showProgress", interactive())]
...
showProgress: 'TRUE' shows progress indicator with estimated time to
          completion for lengthy "by" operations.
```
if showProgress is set to numeric value, we could use that value as the threshold in seconds, instead of the default of 3 seconds.

**Comment by joshhwuu** [comment]
> if showProgress is set to numeric value, we could use that value as the threshold in seconds, instead of the default of 3 seconds.

What about `options(showProgress = ...)`? Would we want to support this with `fread`/`fwrite` as well, as they both read `showProgress` as an option

**Comment by ethanbsmith** [comment]
> you can turn it off using the showProgress argument/option, from ?data.table
> 
that requires hunting down all the scenarios where this is happening and changing code. a global solution would be much preferred

**Comment by joshhwuu** [comment]
> > you can turn it off using the showProgress argument/option, from ?data.table
> 
> that requires hunting down all the scenarios where this is happening and changing code. a global solution would be much preferred

You can use `options=(datatable.showProgress = FALSE)`, it'll set globally

**Comment by ethanbsmith** [comment]
for some reason i thought it was decided against having a global option to govern this. maybe i misread, or something changed. either way, that will solve this for me.

**Comment by ethanbsmith** [comment]
resolved as far as im concerned. your call whether u want to close this.

**Comment by joshhwuu** [comment]
This could be a good feature to think about if more people share the sentiment that a 3s threshold is too little, I don't think it hurts to keep the issue up

**Comment by MichaelChirico** [comment]
The request makes sense to me -- now we just have progress on/off, exposing some more customization is OK. I also worry adding several new options might be overkill, so the suggestion to overload the `datatable.show.progress` argument here makes sense to me. Here are the magic numbers we use for `showProgress=TRUE` now:

https://github.com/Rdatatable/data.table/blob/2a70e2f63cc94c97d35bc493b57321f884539755/src/dogroups.c#L76
https://github.com/Rdatatable/data.table/blob/2a70e2f63cc94c97d35bc493b57321f884539755/src/fwrite.c#L601
https://github.com/Rdatatable/data.table/blob/2a70e2f63cc94c97d35bc493b57321f884539755/src/freadR.c#L681

One downside (upside?) to the change would be forcing the onset time to be the same for fwrite (currently it's 2 where the others are 3).

I also note the extra magic number for fread: IIUC if the initial estimate of progress is >50%, no bar is shown.

**Comment by helske** [comment]
I just noticed this feature. I am bit perplexed by the decision to have option `datatable.showProgress = TRUE` by default in interactive sessions. This means that if `data.table` is used inside a package, users get somewhat cryptic progress messages without possibly even knowing that they are related to `data.table` (and thus they do not necessarily know about `datatable.showProgress` option).

Alternatively, hundreds(?) of package maintainers have to go through all code base to add `showProgress = FALSE` everywhere. Or maybe I am misunderstanding something, given that this feature has existed for some thime already, and I haven't seen any complains?

**Comment by MichaelChirico** [comment]
Makes sense. I still think having progress bars by default in interactive sessions makes sense, but they should definitely get an annotation to clarify they're coming from data.table.

**Comment by helske** [comment]
Would it make sense to have some kind of check on whether data.table is called inside a package namespace, and in such case the progress bar would be off by default? I guess this should be possible but no idea how feasible it would be in practice to check such thing repeatedly.

**Comment by MichaelChirico** [comment]
not worth it IMO. concerned package authors should just turn off progress explicitly.

**Comment by jangorecki** [comment]
> Would it make sense to have some kind of check on whether data.table is called inside a package namespace, and in such case the progress bar would be off by default? I guess this should be possible but no idea how feasible it would be in practice to check such thing repeatedly.

Such hidden behavior will be less clear for majority of package developers, and some of them may actually want to have progress bar. Those will have to dig through our source code/issues to figure out why it works the "smart" way.
It is not too much to expect from package developers to understand how to use their dependencies in a desired way.

**Comment by helmingstay** [comment]
+1 to having a tunable option for update frequency.
 
I'm a long-time data.table user (it's now central to my workflow), and I just recently became aware of this feature. I'm running a series of long jobs on a cluster and at 10-20 min per DT operation, the status updates fill my (already large) tmux buffer. Having a realtime ETA displayed is lovely, and for my use-case a 30-60 sec update window would be much more helpful.

**Comment by jangorecki** [comment]
This message should at least indicate from where it originates. In many cases developer may not even know data.table has been used internally by one of his dependencies and then it is difficult to track to source of it. IMO it could be disabled by default.

Sorry that I repeated what others already said but was just hit by this.
